pkg/registry/rbac/validation/policy_comparator.go

```diff
@@ -171,3 +176,176 @@ func ruleCovers(ownerRule, subRule rbac.PolicyRule) bool {
 }
+
+type collapser struct {
+	rules []rbac.PolicyRule
+}
+
+func (c *collapser) collapseVerb() *collapser {
+	// keyFilter:
+	// NonResourceURLs |Verbs |APIGroups |Resources
+	// --------------------------------------------
+	//              0  |   0  |       1  |       1
+	// appendFilter:
+	// NonResourceURLs |Verbs |APIGroups |Resources
+	// --------------------------------------------
+	//              0  |   1  |       0  |       0
+
+	shouldJump := func(rule rbac.PolicyRule) bool {
+		return len(rule.NonResourceURLs) > 0 || len(rule.Verbs) != 1
+	}
+	c.rules = collapse(c.rules, 3, 4, shouldJump)
+	return c
+}
+func (c *collapser) collapseAPI() *collapser {
+	// keyFilter:
+	// NonResourceURLs |Verbs |APIGroups |Resources
+	// --------------------------------------------
+	//              0  |   1  |       0  |       1
+	// appendFilter:
+	// NonResourceURLs |Verbs |APIGroups |Resources
+	// --------------------------------------------
+	//              0  |   0  |       1  |       0
+
+	shouldJump := func(rule rbac.PolicyRule) bool {
+		return len(rule.NonResourceURLs) > 0 || len(rule.APIGroups) != 1
+	}
+	c.rules = collapse(c.rules, 5, 2, shouldJump)
+	return c
+}
+func (c *collapser) collapseResrouce() *collapser {
+	// keyFilter:
+	// NonResourceURLs |Verbs |APIGroups |Resources
+	// --------------------------------------------
+	//              0  |   1  |       1  |       0
+	// appendFilter:
+	// NonResourceURLs |Verbs |APIGroups |Resources
+	// --------------------------------------------
+	//              0  |   0  |       0  |       1
+
+	shouldJump := func(rule rbac.PolicyRule) bool {
+		return len(rule.NonResourceURLs) > 0 || len(rule.Resources) != 1 || len(rule.ResourceNames) != 0
+	}
+	c.rules = collapse(c.rules, 6, 1, shouldJump)
+	return c
+}
+
+func (c *collapser) collapseNonResourceRules() *collapser {
+	// keyFilter:
+	// NonResourceURLs |Verbs |APIGroups |Resources
+	// --------------------------------------------
+	//              0  |   1  |       0  |       0
+	// appendFilter:
+	// NonResourceURLs |Verbs |APIGroups |Resources
+	// --------------------------------------------
+	//              1  |   0  |       0  |       0
+
+	shouldJump := func(rule rbac.PolicyRule) bool {
+		return len(rule.NonResourceURLs) != 1
+	}
+	c.rules = collapse(c.rules, 4, 8, shouldJump)
+
+	// keyFilter:
+	// NonResourceURLs |Verbs |APIGroups |Resources
+	// --------------------------------------------
+	//              1  |   0  |       0  |       0
+	// appendFilter:
+	// NonResourceURLs |Verbs |APIGroups |Resources
+	// --------------------------------------------
+	//              0  |   1  |       0  |       0
+
+	shouldJump = func(rule rbac.PolicyRule) bool {
+		return len(rule.Verbs) != 1
+	}
+	c.rules = collapse(c.rules, 8, 4, shouldJump)
+
+	return c
+}
+
+func collapse(rules []rbac.PolicyRule, keyFilter, appendFilter int, shouldJump func(rule rbac.PolicyRule) bool) []rbac.PolicyRule {
```

Comment:
this approach is pretty complex to follow. for something so fundamental to the RBAC system, it needs to be much more straightforward. something like https://github.com/liggitt/audit2rbac/blob/master/pkg/util.go#L44 might be an easier approach to follow